src/tests/tribol_tet_mesh.cpp

```diff
@@ -0,0 +1,177 @@
+                                     5., -5. );
+    
+   // sanity checks for this specific mesh
+   EXPECT_EQ( this->m_mesh.mesh_constructed, true );
+   EXPECT_EQ( this->m_mesh.cellType, (int)(tribol::LINEAR_TRIANGLE) );
+   EXPECT_EQ( this->m_mesh.numNodesPerFace, 3 );
+   EXPECT_EQ( this->m_mesh.numNodesPerElement, 4 );
+   EXPECT_EQ( this->m_mesh.numMortarElements, 6 * nMortarElems*nMortarElems*nMortarElems );
+   EXPECT_EQ( this->m_mesh.numNonmortarElements, 6 * nNonmortarElems*nNonmortarElems*nNonmortarElems );
+   EXPECT_EQ( this->m_mesh.numTotalElements, this->m_mesh.numMortarElements + this->m_mesh.numNonmortarElements );
+   EXPECT_EQ( this->m_mesh.numMortarFaces, 2 * nMortarElems*nMortarElems );
+   EXPECT_EQ( this->m_mesh.numNonmortarFaces, 2 * nNonmortarElems*nNonmortarElems );
+   EXPECT_EQ( this->m_mesh.numMortarNodes, (nMortarElems+1)*(nMortarElems+1)*(nMortarElems+1) );
+   EXPECT_EQ( this->m_mesh.numNonmortarNodes, (nNonmortarElems+1)*(nNonmortarElems+1)*(nNonmortarElems+1) );
+
+   //this->m_mesh.testMeshToVtk( "", 1, 1 );
+}
+
+TEST_F( TetMeshTest, build_and_check_mfem_tet_mesh )
+{
+   this->m_mesh.mortarMeshId = 0;
+   this->m_mesh.nonmortarMeshId = 1;
+
+   int nMortarElems = 4; 
+   int nElemsXM = nMortarElems;
+   int nElemsYM = nMortarElems;
+   int nElemsZM = nMortarElems;
+
+   int nNonmortarElems = 3; 
+   int nElemsXS = nNonmortarElems;
+   int nElemsYS = nNonmortarElems;
+   int nElemsZS = nNonmortarElems;
+
+   // mesh bounding box with 0.1 interpenetration gap
+   real x_min1 = 0.;
+   real y_min1 = 0.;
+   real z_min1 = 0.; 
+   real x_max1 = 1.;
+   real y_max1 = 1.;
+   real z_max1 = 1.05;
+
+   real x_min2 = 0.;
+   real y_min2 = 0.;
+   real z_min2 = 0.95;
+   real x_max2 = 1.;
+   real y_max2 = 1.;
+   real z_max2 = 2.;
+
+   this->m_mesh.setupContactMeshTet( nElemsXM, nElemsYM, nElemsZM,
+                                     x_min1, y_min1, z_min1,
+                                     x_max1, y_max1, z_max1,
+                                     nElemsXS, nElemsYS, nElemsZS,
+                                     x_min2, y_min2, z_min2,
+                                     x_max2, y_max2, z_max2,
+                                     5., -5. );
+
+   this->m_mesh.setupMfemMesh();
+
+   // perform mfem mesh based sanity checks
+   EXPECT_EQ( this->m_mesh.mfem_mesh->SpaceDimension(), 3 );
+   EXPECT_EQ( this->m_mesh.mfem_mesh->GetNE(), this->m_mesh.numMortarElements + this->m_mesh.numNonmortarElements );
+   EXPECT_EQ( this->m_mesh.mfem_mesh->GetNV(), this->m_mesh.numMortarNodes + this->m_mesh.numNonmortarNodes );
+   EXPECT_EQ( this->m_mesh.mfem_mesh->GetNFbyType(mfem::FaceType::Boundary), 
+              2 * 6 * (nMortarElems*nMortarElems + nNonmortarElems*nNonmortarElems) );
+
+   // check for inverted elements
+   int wrong_elem_orientation = -1;
+   int wrong_bdr_elem_orientation = -1;
+   bool fix_it = false;
+   wrong_elem_orientation = this->m_mesh.mfem_mesh->CheckElementOrientation( fix_it );
```

Comment:
I see your point here. Even if my TestMesh has inverted elements, let's say, or inverted surface unit normals, `FinalizeTetMesh()` will fix these so this isn't really testing anything. I'll think of another way to test the validity of the mesh. Not sure there is actually a good test for the mfem created mesh, since orientation issues will be fixed. @ebchin - suggestions?

Reply:
Orientation and checks have been updated.